File: app/src/main/java/threading/ex03.java
package threading;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class ex03 {
    public static void main(String[] args) {
        ExecutorService pool = Executors.newFixedThreadPool(3);

        for (int i = 1; i <= 10; i++) {
            int taskNo = i;
            pool.execute(() -> System.out.println(Thread.currentThread().getName() + " task-" + taskNo));
        }

        pool.shutdown();
        try {
            if (!pool.awaitTermination(3, TimeUnit.SECONDS)) {
                pool.shutdownNow();
            }
        } catch (InterruptedException e) {
            pool.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}

File: app-kotlin/src/main/kotlin/threading/ex03.kt
package threading

import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class ex03 {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val pool = Executors.newFixedThreadPool(3)

            for (taskNo in 1..10) {
                pool.execute {
                    println("${Thread.currentThread().name} task-$taskNo")
                }
            }

            pool.shutdown()
            if (!pool.awaitTermination(3, TimeUnit.SECONDS)) {
                pool.shutdownNow()
            }
        }
    }
}
